_posts/2018-02-27-optimizing-grpc-part-1.md

```diff
@@ -0,0 +1,395 @@
+* [KvRunner](https://github.com/carl-mastrangelo/kvstore/blob/01-start/src/main/java/io/grpc/examples/KvRunner.java)
+  orchestrates the interaction between the client and the server.  It is the main entry point,
+  starting both the client and server in process, and waiting for the the client to execute its
+  work.  The runner does work for 60 seconds and then records how many RPCs were completed.
+* [kvstore.proto](https://github.com/carl-mastrangelo/kvstore/blob/01-start/src/main/proto/kvstore.proto)
+  is the protocol buffer definition of our service.  It describes exactly what clients can expect
+  from the service. For the sake of simplicity, we will use Create, Retrieve, Update, and Delete
+  as the operations (commonly known as CRUD).  These operations work with keys and values made up
+  of arbitrary bytes.  While they are somewhat REST like, we reserve the right to diverge and
+  add more complex operations in the future.
+
+While Protocol buffers (protos) aren't a requirement of using gRPC, they are convenient.  We will
+use them to generate the glue code between our application logic, and the core gRPC library.  We
+call this glue code the _stub_.
+
+## Starting Point
+
+### Client
+
+Now that we know what the program _should_ do, we can start looking at how the program performs.
+As mentioned above, the client makes random RPCs.  For example, here is the code that makes the
+[creation](https://github.com/carl-mastrangelo/kvstore/blob/f422b1b6e7c69f8c07f96ed4ddba64757242352c/src/main/java/io/grpc/examples/KvClient.java#L80)
+request:
+
+```java
+  private void doCreate(KeyValueServiceBlockingStub stub) {
+    ByteString key = createRandomKey();
+    try {
+      CreateResponse res = stub.create(
+          CreateRequest.newBuilder()
+              .setKey(key)
+              .setValue(randomBytes(MEAN_VALUE_SIZE))
+              .build());
+      if (!res.equals(CreateResponse.getDefaultInstance())) {
+        throw new RuntimeException("Invalid response");
+      }
+    } catch (StatusRuntimeException e) {
+      if (e.getStatus().getCode() == Code.ALREADY_EXISTS) {
+        knownKeys.remove(key);
+        logger.log(Level.INFO, "Key already existed", e);
+      } else {
+        throw e;
+      }
+    }
+  }
+```
+
+A random key is created, along with a random value.  The request is sent to the server, and the
+client waits for the response.  When the response is returned, the code checks that it is expected,
+and if not, throws an exception.  While the keys are chosen randomly, it is possible that an already
+existing key is in use.  To address this, the code keeps track of keys it has created, so as
+not to create the same key twice.  However, it's possible the key already was created by another
+client, so we log it and move on.  Otherwise, an exception is raised.
+
+The **blocking** gRPC API is being used here, which issues a requests and waits for a response.
```

Comment:
We use the **blocking** gRPC API here, ....

Reply:
Done.